### DISCUSSION TITLE
x-model should accept also a function

### DISCUSSION THREAD

**Discussion by Kilbourne** [alpinejs/alpine#2591] [3. Feature ideas]
Hi!
What about x-model accepting a function? Here a use case:
```
input: {
    ['x-model']() { return this.$store.checkout[this.$el.id] }
}
```

**Comment by SimoTod** [comment]
Doesn't x-bind already support that notation? About your example, x-model needs a string containing the name of the variable, it looks like you are assigning the value instead.

**Comment by Kilbourne** [comment]
Sorry the example should have been like
```
['x-model'](){ return '$store.checkout["'+this.$el.name+'"]'},
```

but the problem remains:
whatever function you give to x-model it's used as string:
```
Alpine Expression Error: Invalid left-hand side in assignment

Expression: "function xModel() {
    return '$store.checkout["' + this.$el.name + '"]';
  } = rightSideOfExpression($event, function xModel() {
    return '$store.checkout["' + this.$el.name + '"]';
  })"

```
because of [https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-model.js#L8-L9](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-model.js#L9)
```
    let assignmentExpression = `${expression} = rightSideOfExpression($event, ${expression})`
```
The expression should be evaluated if is a function

**Reply by SimoTod** [reply]
I see. I'm not sure the PR you mentioned would help because it would do `this[function() {...}]` but it can probably be evaluated at the very top if expression is a function although it might give the false impression that x-model would change over time based on the function outcome.

**Reply by Kilbourne** [reply]
Yes the linked PR is not related. But, as you said ,it could be evaluated once at top. Accessing the element is important to abstract x-model behavior.
Also IMO when using alpine only with extracted components, it's very unintuitive and unexpected that x-model accepts a string but not a function returning string ( unlike most of other directives)

**Reply by Kilbourne** [reply]
Hi @SimoTod , 
I'm trying to override x-model directive with one that accepts function as expression. 
I tried this:
```
import { mutateDom } from 'alpinejs/src/mutation'
import bind from 'alpinejs/src/utils/bind'
import on from 'alpinejs/src/utils/on'

export default (el, { modifiers, expression }, { evaluateLater,evaluate, effect, cleanup }) => {
    expression = typeof expression === 'function' ? evaluate(el,expression) : expression
    let willEvaluate = evaluateLater(el, expression)
    // then i replace evaluate calls with willEvaluate
    //[...same code of x-model...]
```
vs
```
import { evaluateLater } from '../evaluator'
import { directive } from '../directives'
import { mutateDom } from '../mutation'
import bind from '../utils/bind'
import on from '../utils/on'

directive('model', (el, { modifiers, expression }, { effect, cleanup }) => {
    let evaluate = evaluateLater(el, expression)
```
But both `evaluate` and `evaluateLater` get 3 arguments (instead of 2); two times the `el` and as third parameter the `expression`.
Any idea?

**Reply by Kilbourne** [reply]
OK i have seen that evaluateLater,evaluate are already binded to the element in this case, so i removed  el argument from all calls and it works.
I have now an issue with importing the bind utils ( as it is not exposed on Alpine object) 
"(0 , _reactivity__WEBPACK_IMPORTED_MODULE_0__.reactive) is not a function"
